### ISSUE TITLE
USB problem with Windows

### ISSUE THREAD

**Issue by dpgeorge** [micropython/micropython#587]
There have been so far 2 independent reports of a "flaky" serial device on the pyboard.  The pyboard works, and the mass storage device appears.  The serial driver can be installed and the pyboard is recognised by windows, but the terminal program (eg PuTTY) does not work, or at least works only occasionally.  One solution seems to be asserting DTR in the terminal program, if possible.

We need to collect more information about this.

**Comment by bukrat** [comment]
@Deele I tried your way several times and for the first time it worked for a while but after hard reset it disconnected.

**Comment by carminec** [comment]
I all.
I have been having the following problem that might be related to the issue being discussed here.
I used the microPython board for reading three 3-axis accelerometers. A program in C (LabWindows) exploits the serial device connection to the pyboard for sending data request command (print is used to send back ascii strings of data from the board to the PC). The code in the pyboard allows to either process requests for a data read from a single accelerometer , or to initiate a continuous transmission of teh sampled data from the pyboard to the PC (at a rate from 100 to 200 ascii characters per second, depending on the data format). 
Everithing works fine when using an old asus eeePC with Windows 7: the data stream is sent flawlessy form the pyboard to the PC for hours both when using the C program I wrote and when using a terminal emulator (putty). 

However, when I use a brand new XPS 13 DELL computer with windows 8.1, after a few thousands reads (either in send-request/ get-answer mode or continuous transmission mode), the PC apparently stops receiving data (read timeout on the com). Sending data from the PC to the pyboard works and the command interpreter i wrote processes the commands: the pc simply cannot read the data. 
There is no difference when using a Putty terminal in Windows 8.1. The pyboard continues to receive data even after closing the C program (after the timeout error) and using putty to send commands to the command interpreter that is still running in the pyboard, but none of the data being sent back are received. 

In short: no problem whatsoever when using Windows 7; with the very same code and experimental conditions, if I use a Windows 8,1 PC, after a while the pc is no longer capable of receiving the data being sent over the com (but the PC can still send data to the pyboard). 

Has anyone experienced a similar problem? If so, is there a solution? 
Can t be that the serial device driver does not work well with Windows 8? If so, is anyone planning to fix it?   
Thanks.

**Comment by dhylands** [comment]
It sounds very much like there is a bug in the host driver on your Dell machine.

I'm not necessarily ruling out that there isn't a bug on the micropython side (I've been burned enough times in the past by things that seem to be one thing and wind up being something else).

Unfortunately, without being able to reproduce the problem, it would be really hard to debug.

USB is a polling protocol. The host asks the device - do you have anything for me, and the device responds. 
The bug could be of several different varieties:
1 - the host stops asking for data
2 - the device responds that it has no data when in fact it does
3 - Some type of flow control is being triggered (probably accidentally) and that is causing one side to wait for the other side to release the flow control.

I've also investigated issues (not with the pyboard) where the problem is that a faster host can poll the device more quickly (sometimes 1000's of times per second), and that can expose race conditions that aren't seen when using a slower host.

Probably the first thing would be to create some code which you're willing to share which reproduces the problem, and which other users can try (I have mostly linux machines, although I do have access to a few Windows machines). I'll have to check if any of them are running Windows 8 (my own personal Windows machines are all XP or Win 7).

**Comment by carminec** [comment]
@dpgeorge 
Thank you very much for your prompt replay.
I will prepare a simplified version of the command interpreter running in my micropython so that the issue can be tested using putty (I'm new with python and my code, at the moment, is quite messy although it seems to work fine with Windows 7). Before doing that, however, I will test my project with another Windows 8.1 system  (running on a lenovo youga 2 low end notebook) an I will let you know the results.  
Best regards,

    Carmine.

**Comment by carminec** [comment]
@dhylands
Hi Dave.
First, I would like to apologize for not directing to you my replay to your post: I must have messed up while copying and pasting nicknames.....
Second, it appears that your first guess was the correct one: I experience the problem I described in 
my earlier post only on my new "superduper" Dell XPS 13...... :-(
Everything works fine with other Windows 8.1 based systems or high-end Windows 7 based laptops.
I have also rewritten a small command interpreter for testing this problem using putty and, should you think it may be useful to others, I can certainly share it (as soon as I learn how to do it). 
Thank you very much for your help.
Sincerely, 

     Carmine.

**Comment by dhylands** [comment]
For small text things, you can share them using https://gist.github.com/

I often use Dropbox, or you could create a repository on github and check the files into that.

I also often use tools like pastebin: http://pastebin.com/

**Comment by msiemens** [comment]
Maybe it's a problem with a Dell driver? I'm still having these issues and I'm using a a Dell Precision T3610.

**Comment by stinos** [comment]
Maybe a bit far fetched, but: I've had USB problems on T3610 and similar but only when using a USB 3 port where using one of the non-ss ports for the same device was ok. From the problem descriptions here I doubt it's that, but you never know..

**Comment by msiemens** [comment]
@stinos: I've already tried using a non-SS port but it didn't make any difference.

I've found another report of a similar problem: Dell chipset, flaky connection with the serial USB driver. He's experiencing problems when using the device through a USB 2.0 fullspeed hub. (http://www.winvistatips.com/threads/bug-in-usbser-sys-when-using-a-high-speed-hub.194444/)

**Comment by kdschlosser** [comment]
I am not sure if this would be a related issue.
I have a knock off Wemos D1 min using the CH340 USB/Serial
Window 7 x64 SP1.
Tried using Putty, TeraTerm, RealTerm Nothing seems to work

Wemos erases properly (no errors), flashes properly (no errors)
I can see the ESP AP
But when i connect it to via serial the REPL prompt comes up and then I can't do a thing. pressing the reset button i get a new REPL prompt and stil can't do a thing

It seems as tho it's not accepting any keyboard input.

I have another ESP the adafruit Huzzah, it uses FTDI USB/Serial and it works just fine. No issues at all.

Is there a way to get WebREPL running and get it out of AP mode and connected to my Wifi without having to use the Serial and manual entry using the keyboard?

**Comment by kdschlosser** [comment]
well I have gotten it to work. and maybe something like this might for other if they experience the same issue,

i flashed the wemos D1 (knockoff) with the Arduino IDE and then flashed it again with micropython. and now it works properly. I am not sure what the Arduino IDE could have done to make micropython work. but if someone else can test this and see if it solves the issue of not being able to input anything into the REPL then it would be worth an investigation

**Comment by METASHONGKONG** [comment]
I have also encountered your problem... I have many 8266. For some 8266, it works both serial REPL and wifi REPL. Some only works for wifi REPL but not works for serial REPL. I cannot type anything in serial REPL. If it works for both serial and wifi, it would be great. However, just few (2 pieces in 10) of the 8266 work for both.

BTW, how do you flash wemos D1 via Adruino IDE? I just use lua and python for 8266 before only.

Thanks.

**Comment by kdschlosser** [comment]
@METASHONGKONG

The link below will provide the software and instruction on how to flash it 
all ya have to do is to flash it with one of the example sketches and then flash micropython afterwards. This worked for me.. tho i was only able to do it with a single ESP that was having the problem. so I have not been able to confirm 100% that this is a solution.. but i did flash the ESP a whole bunch of times with micropython and it would still not work.. When i flashed it with the arduino IDE and then again with micropython is worked.


https://github.com/esp8266/Arduino

**Comment by juransaber** [comment]
I also have this problem on nodemcu with v1.9.4, usb chip is cp2102, but another nodemcu board which use ch340 works well
I'm using mac osx, screen to connect

**Comment by dobotopensource** [comment]
If the DTR in the xcomV2.0 is not checked, the condition blow won't be satisfied!!
`if (cdc->connect_state != USBD_CDC_CONNECT_STATE_DISCONNECTED)`

**Comment by tlegrain** [comment]
Unable to use serial terminal REPL on Windows 10. Never worked
- Received Pyboard this week.
- PYBFLASH drive always works. Pyboard function, can program with main.py in flash drive.
- Windows 10 Pro 10.0.17763
- Lenovo ThinkPad P50
- Proper USB Driver installed. Had to forced un-signed USB Driver, pybcdc.inf. (Ports (COM&LPT) = Pyboard USB Comm Port (COM20) 
- Pybv1.1
- DFUed: pybv11-20190713-v1.11-146-g154062d9c.dfu. Note that it also did not work with original firmware.
- Tested with two different USB cables. 
- Using a powered USB hub
- Using directly connected to PC (on its two USB connector).
- Used PuTTY, TeraTerm (setdtr 1, and setdtr 0), Br@y's Terminal (DTR on and DTR off).

**Comment by kdschlosser** [comment]
you can try what I did to get your board working. it is a few posts back.

**Comment by chrismas9** [comment]
There is no need to install the driver for Windows 10 for Pyboard or any STM32 board with USB connected directly to the STM32. I use Pyboard, NUCLEO, and proprietary boards with all versions of Windows 10.

Since DFU worked you probably don't have a hardware or cable problem. Try plugging the Pyboard into another PC and open Device Manager. It should show under Ports. If that works you may need to remove the driver you installed - it's only for Windows 7.

**Comment by tlegrain** [comment]
Thank you for your input.
- originally tried with the default Windows 10 driver. No serial (ie. not REPL).
- I too have used STM32 Discovery and Nucleo without issues on Windows 10.

- Using a Windows 7 machine:
- Plugged in PyBoard
- Installing drivers ...
- Recognized PYBFLASH drive
- Error: Device Driver was not succesfully installed
- Pyboard Virtual Comm Port in FS Mode No Driver Found
- Device manager shows Other devices with !Pyboard Virtual Comm Port in FS Mode
- Updated driver with pybcdc.inf successful
- Putty or TeraTerm still do not function via REPL

**Comment by kdschlosser** [comment]
I do want to mention something..

These ESP chips are extremely voltage sensitive. If you are using a USB port that does not output a high enough current you can be ending up with an issue because of that. The esp chips will draw damned close to 500ma when flashing them. and that is not counting efficiency loss through cabling, connections, regulators... You may want to locate yourself a good wall wart os an old PC power supply that you can "hot wire" to the board. have you tried plugging the thing into a different USB root hub? Typically a mainboard on a PC will have groupings of 4 USB ports to one root hub. and all 4 can sometimes share the same power supply. so if you have some other devices plugged into the same hub and they are using up enough of the current this will cause you flashing problems.

I am also going to still suggest you flash the board using the Arduino ESP library. once you do that then flash it with micropython. I do not know why that has worked for me on more then one occasion. But it has. It is not hard to do and will only take a few minutes of your time,

**Comment by kdschlosser** [comment]
I also just noticed that someone mentioned unsigned drivers in Windows 10.

If you are running Windows 10 and the drivers appear to have installed correctly and there is no indication of any kind of an error. But things are not working like they should. In the link below is a tutorial i wrote on how to get Windows 10 into an unsigned driver mode without having Windows set into "test mode" (no water mark on the lower right of the desktop). 

http://www.eventghost.net/forum/viewtopic.php?f=10&t=9658

The directions above have been proven to work without issue for quite a few Windows 10 driver ailments. It is worth giving it a shot. The people over there at Microsoft thought it would be funny to disallow unsigned drivers and set the whole thing up so that it appears as tho the driver installed and is functioning properly and give you no message or error telling you that it is not. Because of this Users have found themselves getting really aggravated because everything looks like it should be functioning as it should be. 

If anyone follows the above directions let us know the outcome.

When dealing with windows and serial ports it is similiar to NIX based handling and it is also different at the same time. Nix based systems is based on a file that represents the port. data gets written to and read from that file. the windows process is similiar to that.. except for one component... there is no file. It is a database namespace entry. So the process of pressing the reset button causes the connection to get released from the device end. anything connected on the other end is going to have a timeout period. depending on the application this timeout can take a while. point in fact is putty. putty takes a long while to see the drop. until that drop is seen by the application the application will remain connected to that database record. Windows is unable to remove the record because it is in use. and because the record cannot be removed a new database entry cannot be created when the device "reattaches" because it is trying to create a record that has the same port name. there are several things that you can try and might solve the this issue. one is adjust the timeout duration in the application. the second is to slow down the reset process of the device allowing for the application to timeout and disconnect from the database record. thus allowing a new record to be created.

**Comment by jonnor** [comment]
Not a reproducible problem, and over 5 year since last comment. Perhaps this issue can be closed? And then people should open new, specific and detailed issues if they have problems.

**Comment by Metallicow** [comment]
Ummm, haven't had contact with some the kickstarter boards in a few years. I Recall there was a restarting issue... This may be windows specific, (some people think it is a dongle... (obviously NOT my case)(RAW Kickstarter board), 
Some flirted the idea that it might be a firmware change since then...(0,1) ,
And some folks like me are just crazy.... That is true after 5+ years. Thank you Damian.

Anyhow, I would like the project to continue without hiccups, I am still here if you need to hollar, Thanks.
